src/plugins/internalnotification/internalnotification.c

```diff
@@ -170,167 +234,136 @@ void elektraInternalnotificationUpdateRegisteredKeys (Plugin * plugin, KeySet *
+			{
+				changed = 1;
+				key = checkKey;
+			}
+			else
+			{
+				keyDel (checkKey);
+			}
 		}
 		else
 		{
-			const char * currentValue = keyString (key);
-			changed = registeredKey->lastValue == NULL || strcmp (currentValue, registeredKey->lastValue) != 0;
+			key = ksLookupByName (keySet, registeredKey->name, 0);
+			if (key == NULL)
+			{
+				registeredKey = registeredKey->next;
+				continue;
+			}
 
-			if (changed)
+			// Detect changes for string keys
+			if (!keyIsString (key))
+			{
+				// always notify for binary keys
+				changed = 1;
+			}
+			else
 			{
-				// Save last value
-				char * buffer = elektraStrDup (currentValue);
-				if (buffer)
+				const char * currentValue = keyString (key);
+				changed = registeredKey->lastValue == NULL || strcmp (currentValue, registeredKey->lastValue) != 0;
+
+				if (changed)
 				{
-					if (registeredKey->lastValue != NULL)
+					// Save last value
+					char * buffer = elektraStrDup (currentValue);
+					if (buffer)
 					{
-						// Free previous value
-						elektraFree (registeredKey->lastValue);
+						if (registeredKey->lastValue != NULL)
+						{
+							// Free previous value
+							elektraFree (registeredKey->lastValue);
+						}
+						registeredKey->lastValue = buffer;
 					}
-					registeredKey->lastValue = buffer;
 				}
 			}
 		}
 
 		if (changed)
 		{
-			// Perform actions depending on type
-			switch (registeredKey->type)
-			{
-			case TYPE_INT:
-				ELEKTRA_LOG_DEBUG ("found registeredKey=%s; updating variable=%p with string value \"%s\"",
-						   registeredKey->name, (void *) registeredKey->ref.variable, keyString (key));
-
-				// Convert string value to long
-				char * end;
-				errno = 0;
-				long int value = strtol (keyString (key), &end, 10);
-				// Update variable if conversion was successful and did not exceed integer range
-				if (*end == 0 && errno == 0 && value <= INT_MAX && value >= INT_MIN)
-				{
-					*(registeredKey->ref.variable) = value;
-				}
-				else
-				{
-					ELEKTRA_LOG_WARNING ("conversion failed! keyString=\"%s\" *end=%c, errno=%d, value=%ld",
-							     keyString (key), *end, errno, value);
-				}
-				break;
-			case TYPE_CALLBACK:
-				ELEKTRA_LOG_DEBUG ("found registeredKey=%s; invoking callback", registeredKey->name);
-				ElektraNotificationChangeCallback callback =
-					*(ElektraNotificationChangeCallback) registeredKey->ref.callback;
-				callback (key);
-				break;
-			}
+			ELEKTRA_LOG_DEBUG ("found changed registeredKey=%s with string value \"%s\". using context or variable=%p",
+					   registeredKey->name, keyString (key), registeredKey->context);
+
+			// Invoke callback
+			ElektraNotificationChangeCallback callback = *(ElektraNotificationChangeCallback) registeredKey->callback;
+			callback (key, registeredKey->context);
 		}
 
 		registeredKey = registeredKey->next;
 	}
 }
 
+// Generate register and conversion functions
+INTERNALNOTIFICATION_TYPE (int, long int, Int, (strtol (string, &end, 10)),
```

Comment:
Very compact, well done!

Is this somehow reusable for other places? (maybe as [super macro](http://s11n.net/papers/supermacros_cpp.pdf)?)